server/src/main/java/io/crate/planner/statement/CopyFromPlan.java

```diff
@@ -288,23 +293,63 @@ public static ExecutionPlan planCopyFromExecution(AnalyzedCopyFrom copyFrom,
                 lineNumberSymbol
             );
         } else {
-            sourceIndexWriterProjection = new SourceIndexWriterProjection(
+
+            // Those columns must be collected regardless of their presence in targetColumns
+            // since they are used for calculating shard id.
+            // Collecting all top-level table columns might miss those columns if they are sub-columns.
+            Set<Reference> targetColumns = new HashSet<>();
+            targetColumns.addAll(primaryKeyRefs);
+            targetColumns.addAll(table.partitionedByColumns());
+            if (clusteredBy != null) {
+                targetColumns.add(table.getReference(clusteredBy));
+            }
```

Comment:
Not reusing `getSymbolsRequiredForShardIdCalc` because it works with symbols (unwrap generated reference which is not needed for the Indexer code path) and also I wanted to have Set to adding top-level columns without duplicates. 

Anyway, `getSymbolsRequiredForShardIdCalc` will be gone in the next iteration, when RETUEN SUMMARY will be switched to no-raw.